packages/picasso.js/src/web/renderer/canvas-renderer/shapes/image.js

```diff
@@ -0,0 +1,106 @@
+let offscreenBuffer = null;
+function loadImage(src, onLoad) {
+  const image = new Image();
+  image.src = src;
+
+  image.onload = () => {
+    onLoad(image);
+  };
+
+  image.onerror = (e) => {
+    console.error('Image load error', e);
+  };
+}
+export function positionImage(img) {
+  const position = img.imgPosition || 'center-center';
+  if (img.symbol === 'circle') {
+    const radius = Math.min(img.width, img.height) / 2;
+    img.width = radius * 2;
+    img.height = radius * 2;
+  }
+  if (position) {
+    switch (position) {
+      case 'top-center':
+        img.y -= img.height / 2;
+        break;
+      case 'center-left':
+        img.x -= img.width / 2;
+        break;
+      case 'center-right':
+        img.x += img.width / 2;
+        break;
+      case 'top-left':
+        img.x -= img.width / 2;
+        img.y -= img.height / 2;
+        break;
+      case 'top-right':
+        img.x += img.width / 2;
+        img.y -= img.height / 2;
+        break;
+      case 'bottom-left':
+        img.x -= img.width / 2;
+        img.y += img.height / 2;
+        break;
+      case 'bottom-right':
+        img.x += img.width / 2;
+        img.y += img.height / 2;
+        break;
+      case 'bottom-center':
+        img.y += img.height / 2;
+        break;
+      default:
+        break;
+    }
+  }
+}
+export default function render(img, { g }) {
+  const canvas = g.canvas;
+
+  const ratio = window.devicePixelRatio || 1;
+  const logicalWidth = canvas.clientWidth;
+  const logicalHeight = canvas.clientHeight;
+
+  const renderWidth = logicalWidth * ratio;
+  const renderHeight = logicalHeight * ratio;
+
+  if (!offscreenBuffer || offscreenBuffer.width !== renderWidth || offscreenBuffer.height !== renderHeight) {
+    offscreenBuffer = document.createElement('canvas');
```

Comment:
Can the `canvas` element be created outside the scope of the `render` function so that all calls to render a image shares the same canvas? Seems like a small but possible optimazation